src/importer.cpp

```diff
@@ -0,0 +1,483 @@
+ * the implementation from the definition allows for greater flexibility when
+ * distributing the code.
+ */
+struct Importer::ImporterImpl
+{
+    Importer *mImporter = nullptr;
+
+    ImportLibrary mLibrary;
+    std::vector<std::string> mLibraryIndex;
+    std::vector<std::pair<std::string, std::string>> mExternals;
+
+    bool resolveImport(const ImportedEntityPtr &importedEntity,
+                       const std::string &destination,
+                       const std::string &typeString,
+                       const std::string &baseFile,
+                       std::vector<std::tuple<std::string, std::string, std::string>> &history);
+
+    bool resolveComponentImports(const ComponentEntityPtr &parentComponentEntity,
+                                 const std::string &baseFile,
+                                 std::vector<std::tuple<std::string, std::string, std::string>> &history);
+
+    bool doResolveImports(ModelPtr &model, const std::string &baseFile,
+                          std::vector<std::tuple<std::string, std::string, std::string>> &history);
+};
+
+Importer::Importer()
+    : mPimpl(new ImporterImpl())
+{
+    mPimpl->mImporter = this;
+}
+
+ImporterPtr Importer::create() noexcept
+{
+    return std::shared_ptr<Importer> {new Importer {}};
+}
+
+Importer::~Importer()
+{
+    delete mPimpl;
+}
+
+/**
+ * @brief Resolve the path of the given filename using the given base.
+ *
+ * Resolves the full path to the given @p filename using the @p base.
+ *
+ * This function is only intended to work with local files.  It may not
+ * work with bases that use the 'file://' prefix.
+ *
+ * @param filename The @c std::string relative path from the base path.
+ * @param base The @c std::string location on local disk for determining the full path from.
+ *
+ * @return The full path from the @p base location to the @p filename
+ */
+std::string resolvePath(const std::string &filename, const std::string &base)
+{
+    // We can be naive here as we know what we are dealing with.
+    std::string path = base.substr(0, base.find_last_of('/') + 1) + filename;
+    return path;
+}
+
+bool checkUnitsForCycles(ModelPtr &model, std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    // Check through model for overlap with the current history so we can report cycles properly.
+    for (size_t u = 0; u < model->unitsCount(); ++u) {
+        auto units = model->units(u);
+        if (units->isImport()) {
+            ImportSourcePtr importSource = units->importSource();
+            auto h = std::make_tuple(units->name(), units->importReference(), importSource->url());
+            if (std::find(history.begin(), history.end(), h) != history.end()) {
+                history.emplace_back(h);
+                return false;
+            }
+        }
+    }
+    return true;
+}
+bool checkComponentForCycles(ModelPtr &model, std::vector<std::tuple<std::string, std::string, std::string>> &history)
```

Comment:
Insert a blank line between 121 and 122.

Reply:
7b7ad654